File: 2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        int mfb[] = new int[nums.length];
        int maxi= Integer.MIN_VALUE;
        for(int i=nums.length-1; i>=0; i--){
            maxi= Math.max(nums[i], maxi);
            mfb[i]= maxi;
        }
        long ans=0;
        maxi = nums[0];
        for(int i=1; i<nums.length-1; i++){
            ans= Math.max(ans, 1L*(maxi-nums[i])*mfb[i+1]);
            maxi= Math.max(maxi, nums[i]);
        }
        return ans;
    }
}

File: 2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.cpp
class Solution {
public:
#define ll long long int
    long long maximumTripletValue(vector<int>& nums) {
        map<int,int> mp;
        for(int i=0; i<nums.size()-1; i++) mp[nums[i]]++;
        ll k= nums.back(), ans=0;
        for(int i=nums.size()-2; i>0; i--){
            mp[nums[i]]--;
            if(mp[nums[i]]<=0){
                mp.erase(nums[i]);
            }
            ll jth= (*(--mp.end())).first;
            ans = max(ans, (jth- nums[i]) * k);
            k= max(k, 1LL*nums[i]);
        } 
        return ans;
    }
    
};